### PR TITLE
Added support for negative reaction-orders as well as non-reactant or…

### PR DESCRIPTION
**Pull request by santoshshanbhogue**
…ders. In addition checked that the changes are compatible with duplicate reactions. So reactions can have multiple option strings.

### CODE REVIEW

**Comment by santoshshanbhogue** [comment]
The premixed twin flame example does not work anymore in 2.3.0a2 with GRI3.0. It works with the Smooke mechanism...

The main thing here was the method of computing the strain rate. I corrected that. Plus added a function to compute the consumption speed, since this is a popular use for the twin flame solver. For this I had to use the scipy.integrate module

Can I upload a Jupyter notebook directly ? That would be super-illustrative... Perhaps in a new directory called visualExamples ?

File: interfaces/cython/cantera/examples/onedim/premixed_counterflow_twin_flame.py [hunk 1]
```diff
@@ -1,20 +1,59 @@
...
-simulation differs from the similar premixed_counterflow_flame.py example as the
-latter simulates a jet of reactants shooting into products.
-"""
-
 import cantera as ct
 import numpy as np
 
-# This function is called to run the solver
+#Trapz is used solely to compute consumption speed
+from scipy.integrate import trapz
```
Comment (speth) [inline interfaces/cython/cantera/examples/onedim/premixed_counterflow_twin_flame.py]: `np.trapz` is the same function

File: interfaces/cython/cantera/examples/onedim/premixed_counterflow_twin_flame.py [hunk 2]
```diff
@@ -1,20 +1,59 @@
...
+    return Sc
+
+# This function is called to run the solver                                                                           
 def solveOpposedFlame(oppFlame, massFlux=0.12, loglevel=1,
-                      ratio=3, slope=0.15, curve=0.25, prune=0.05):
-    """
-    Execute this function to run the Oppposed Flow Simulation This function
-    takes a CounterFlowTwinPremixedFlame object as the first argument
+                      ratio=2, slope=0.3, curve=0.3, prune=0.05):
+    """
```
Comment (speth) [inline interfaces/cython/cantera/examples/onedim/premixed_counterflow_twin_flame.py]: Please avoid introducing trailing whitespace

**Comment by speth** [comment]
The bug report for the premixed counterflow twin flame example is appreciated, but it would be better to create a new issue for something unrelated to this PR.

I would also recommend creating a new branch in Git for each PR, so you don't end up with additional, unrelated commits in your PR when you make more commits to master.

**Comment by santoshshanbhogue** [comment]
Thanks Ray for the pointers. I have created a separate branch for the twin-flame changes

**Comment by santoshshanbhogue** [comment]
I assume you can reject commit 1a26baa on this branch

**Comment by speth** [comment]
I think you can (and should) implement this without introducing any new global variables in `ck2cti.py`. The flags read from the command line in `main()` can be passed as additional arguments to `parser.convertMech`. This will also allow you to introduce some test cases of this feature in `test_convert.py`.

File: interfaces/cython/cantera/ck2cti.py [hunk 3]
```diff
@@ -312,14 +316,55 @@ def to_cti(self, indent=0):
...
+            if 'negative_orders' in options:
+                kinstr = kinstr[:-1] + ",\n{0}order='{1}', options={2})".format(k_indent, order, optStr)
+            else:
+                kinstr = kinstr[:-1] + ",\n{0}order='{1}')".format(k_indent, order)
+
+        if self.duplicate and 'negative_orders' not in options:
+            kinstr = kinstr[:-1] + ",\n{0}options={1})".format(k_indent, optStr)
+            
+                
+        if len(options) == 1:
```
Comment (speth) [inline interfaces/cython/cantera/ck2cti.py]: I think the logic for creating `optStr` needs to be needs to be re-examined. In the loops above, you can simply update `options`, then `optStr` can be generated just once. Then, the condition for appending an `options=...` section to `kinStr` should be conditioned simply on `options` being non-empty (i.e. just `if options`: ...).

File: interfaces/cython/cantera/ck2cti.py [hunk 4]
```diff
@@ -312,14 +316,55 @@ def to_cti(self, indent=0):
...
             order = ' '.join('{0}:{1}'.format(k,v)
                              for (k,v) in self.fwdOrders.items())
-            kinstr = kinstr[:-1] + ",\n{0}order='{1}')".format(k_indent, order)
 
+            for(k,v) in self.fwdOrders.items():
+                #Check for negative orders
+                if float(v) < 0 and NEGATIVE_ORDERS is False:
+                    #Meaning negative order detected but --allow-negative-order flag not passed to ck2cti
+                    print('\nWARNING: Reaction with negative order detected.') 
+                    print('Re-execute ck2cti with --allow-negative-orders and also execute ck2cti --help to see note on danger of negative orders.\n')
```
Comment (speth) [inline interfaces/cython/cantera/ck2cti.py]: I am concerned that this warning does not sufficiently dissuade use of this flag -- it just encourages the user to blindly re-run `ck2cti` with the option enabled.

Comment (santoshshanbhogue) [inline interfaces/cython/cantera/ck2cti.py]: Can't think of an alternative. The goal is to get some mechanisms that hitherto do not work with Cantera to work, right ?

**Pull request closed without merging**